caom2pipe/client_composable.py

```diff
@@ -0,0 +1,665 @@
+    end = current()
+    file_size = os.stat(fqn).st_size
+    metrics.observe(start, end, file_size, 'get', 'data', file_name)
+
+
+def data_put_fqn(
+    client,
+    source_name,
+    storage_name,
+    stream='raw',
+    metrics=None,
+):
+    """
+    Make a copy of a locally available file by writing it to CADC. Assumes
+    file and directory locations are correct. Requires a checksum comparison
+    by the client.
+
+    :param client: The CadcDataClient for write access to CADC storage.
+    :param source_name: str fully-qualified
+    :param storage_name: StorageName instance
+    :param stream: str A relic of the old CADC storage.
+    :param metrics: Tracking success execution times, and failure counts.
+    """
+    start = current()
+    try:
+        client.put_file(
+            storage_name.archive,
+            source_name,
+            archive_stream=stream,
+            mime_type=storage_name.mime_type,
+            mime_encoding=storage_name.mime_encoding,
+            md5_check=True,
+        )
+        file_size = os.stat(source_name).st_size
+    except Exception as e:
+        metrics.observe_failure('put', 'data', source_name)
+        logging.debug(traceback.format_exc())
+        raise mc.CadcException(f'Failed to store data with {e}')
+    end = current()
+    metrics.observe(start, end, file_size, 'put', 'data', source_name)
+
+
+def declare_client(config):
+    """Common code to set the client used for interacting with CADC
+    storage."""
+    if config.features.supports_latest_client:
+        logging.warning('Using vos.Client for storage.')
+        cert_file = config.proxy_fqn
+        if cert_file is not None and os.path.exists(cert_file):
+            cadc_client = Client(vospace_certfile=cert_file)
+        else:
+            raise mc.CadcException(
+                'No credentials configured or found. Stopping.'
+            )
+    else:
+        logging.warning('Using cadcdata.CadcDataClient for storage.')
+        subject = define_subject(config)
+        cadc_client = CadcDataClient(subject)
+    return cadc_client
+
+
+def define_subject(config):
+    """Common code to figure out which credentials to use based on the
+    content of a Config instance."""
+    subject = None
+    if config.proxy_fqn is not None and os.path.exists(config.proxy_fqn):
+        logging.debug(
+            f'Using proxy certificate {config.proxy_fqn} for credentials.'
+        )
+        subject = net.Subject(username=None, certificate=config.proxy_fqn)
+    elif config.netrc_file is not None:
+        netrc_fqn = os.path.join(config.working_directory, config.netrc_file)
+        if os.path.exists(netrc_fqn):
+            logging.debug(f'Using netrc file {netrc_fqn} for credentials.')
+            subject = net.Subject(
+                username=None, certificate=None, netrc=netrc_fqn
+            )
+        else:
+            logging.warning(f'Cannot find netrc file {netrc_fqn}')
```

Comment:
The pipelines can be run in circumstances where no credentials are required, so that's why it's only a warning message. For example, the pipelines have a 'SCRAPE' task type, for working off-line, and testing pipeline development.

w.r.t. `.netrc` files no longer being supported - the `omm2caom2` pipeline, as far as I know, is the only user of `.netrc` files. It will be a large leap for Sylvie to change A&A mechanisms, so it would be nice if this change is not made abruptly.